### PR TITLE
quick fingerprinting tool

### PR DESCRIPTION
**Pull request by doron-st**
detection of sample_id from cram by calculating snv recall on a certain genomic window

### CODE REVIEW

**Comment by doron-st** [comment]
```Summary of errors:

ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V125/039272-20240513_0500/039272-NA24149-Z0115-CGGCTAGATGCAGAT/039272-NA24149-Z0115-CGGCTAGATGCAGAT.cram - HG003 does not match ground truth hit_fraction=0.5516111238350845, hit_count=171
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V125/039272-20240513_0500/039272-NA24149-Z0115-CGGCTAGATGCAGAT/039272-NA24149-Z0115-CGGCTAGATGCAGAT.cram - HG003 matched ground truth of HG005, hit_fraction=0.9999969040343529, count=323
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V125/039272-20240513_0500/039272-NA24149-Z0115-CGGCTAGATGCAGAT/039272-NA24149-Z0115-CGGCTAGATGCAGAT.cram - HG003 does not have a maximal match with it's expected ground truth hit_fraction=0.5516111238
350845, max_hit_fraction = 0.9999969040343529
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V140/034338-20240514_1203/034338-NA24149-Z0115-CGGCTAGATGCAGAT/034338-NA24149-Z0115-CGGCTAGATGCAGAT.cram - HG003 does not match ground truthhit_fraction=0.5548369198809037, hit_count=172
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V140/034338-20240514_1203/034338-NA24149-Z0115-CGGCTAGATGCAGAT/034338-NA24149-Z0115-CGGCTAGATGCAGAT.cram - HG003 matched ground truth of HG005, hit_fraction=0.9999969040343529, count=323
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V140/034338-20240514_1203/034338-NA24149-Z0115-CGGCTAGATGCAGAT/034338-NA24149-Z0115-CGGCTAGATGCAGAT.cram - HG003 does not have a maximal match with it's expected ground truthhit_fraction=0.5548369198
809037, max_hit_fraction = 0.9999969040343529
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V125/039272-20240513_0500/039272-NA24143-Z0016-CATCCTGTGCGCATGAT/039272-NA24143-Z0016-CATCCTGTGCGCATGAT.cram - HG004 does not match ground truthhit_fraction=0.6698996606812588, hit_count=138
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V125/039272-20240513_0500/039272-NA24143-Z0016-CATCCTGTGCGCATGAT/039272-NA24143-Z0016-CATCCTGTGCGCATGAT.cram - HG004 matched ground truth of HG006, hit_fraction=0.9999961390110463, count=259
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V125/039272-20240513_0500/039272-NA24143-Z0016-CATCCTGTGCGCATGAT/039272-NA24143-Z0016-CATCCTGTGCGCATGAT.cram - HG004 does not have a maximal match with it's expected ground truthhit_fraction=0.669899
6606812588, max_hit_fraction = 0.9999961390110463
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V125/039272-20240513_0500/039272-NA24631-Z0024-CTGAGCCTGTCAGAT/039272-NA24631-Z0024-CTGAGCCTGTCAGAT.cram - HG005 does not match ground truthhit_fraction=0.6687285797876787, hit_count=216
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V125/039272-20240513_0500/039272-NA24631-Z0024-CTGAGCCTGTCAGAT/039272-NA24631-Z0024-CTGAGCCTGTCAGAT.cram - HG005 matched ground truth of HG006, hit_fraction=0.9999961390110463, count=259
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V125/039272-20240513_0500/039272-NA24631-Z0024-CTGAGCCTGTCAGAT/039272-NA24631-Z0024-CTGAGCCTGTCAGAT.cram - HG005 does not have a maximal match with it's expected ground truthhit_fraction=0.6687285797
876787, max_hit_fraction = 0.9999961390110463
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V140/034338-20240514_1203/034338-NA24631-Z0024-CTGAGCCTGTCAGAT/034338-NA24631-Z0024-CTGAGCCTGTCAGAT.cram - HG005 does not match ground truthhit_fraction=0.6656326141405136, hit_count=215
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V140/034338-20240514_1203/034338-NA24631-Z0024-CTGAGCCTGTCAGAT/034338-NA24631-Z0024-CTGAGCCTGTCAGAT.cram - HG005 matched ground truth of HG006, hit_fraction=0.9999961390110463, count=259
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V140/034338-20240514_1203/034338-NA24631-Z0024-CTGAGCCTGTCAGAT/034338-NA24631-Z0024-CTGAGCCTGTCAGAT.cram - HG005 does not have a maximal match with it's expected ground truthhit_fraction=0.6656326141
405136, max_hit_fraction = 0.9999961390110463```

**Comment by doron-st** [comment]
Improved error message to be oneliner:
```s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V125/039272-20240513_0500/039272-NA24149-Z0115-CGGCTAGATGCAGAT/039272-NA24149-Z0115-CGGCTAGATGCAGAT.cram - HG003 vs. HG007 hit_fraction=0.6740561550861307
ERROR: s3://ultimagen-pipelines-503496316425-us-east-1-runs-data/V125/039272-20240513_0500/039272-NA24149-Z0115-CGGCTAGATGCAGAT/039272-NA24149-Z0115-CGGCTAGATGCAGAT.cram - HG003 does not match it's ground truth: hit_fraction=0.5516111238350845 matched ground truth of HG005: hit_fraction=0.9999969040343529 does not have a maximal match with it's expected ground truth: vimax_hit_fraction = 0.9999969040343529```

File: ugvc/validation/variant_hit_fraction_caller.py
Comment (ilyasoifer) [inline ugvc/validation/variant_hit_fraction_caller.py:1]: maybe belongs to comparison? I am not sure what "validation" stands for

File: ugvc/validation/variant_hit_fraction_caller.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        self.out_dir = out_dir
+        self.sp = sp
+        self.min_af_snps = min_af_snps
+
+    def call_variants(self, cram: str, vcf: str, regions: str, min_af: float) -> None:
+        self.sp.print_and_run(
+            f"bcftools mpileup {cram} -f {self.ref} -T {regions} -a ad,format/dp --skip-indels -d 500 "
+            + f"| bcftools view -i 'AD[0:1] / format/DP >= {min_af}' -Oz -o {vcf}"
+        )
+        self.sp.print_and_run(f"bcftools index -t {vcf}")
```
Comment (ilyasoifer) [inline ugvc/validation/variant_hit_fraction_caller.py]: better to use vcf_pipeline_utils.index_vcf

File: ugvc/validation/variant_hit_fraction_caller.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        with open(f"{self.out_dir}/{cram_base_name}_{gt_base_name}.hit.txt", "w", encoding="utf-8") as fh:
+            fh.write(f"hit_count {hit_count}\n")
+            fh.write(f"hit_fraction {hit_fraction}\n")
+
+        return hit_fraction, hit_count, ground_truth_count
+
+    def count_lines(self, in_file: str, out_file: str):
+        self.sp.print_and_run(f"wc -l {in_file} " + "| awk '{print $1}' " + f" > {out_file}")
+
+    def vcf_to_bed(self, vcf: str, bed: str, max_vars=10**8, region: str = "") -> None:
```
Comment (ilyasoifer) [inline ugvc/validation/variant_hit_fraction_caller.py]: Maybe not awk? Something a little less hacky?

File: ugvc/validation/variant_hit_fraction_caller.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    def call_variants(self, cram: str, vcf: str, regions: str, min_af: float) -> None:
+        self.sp.print_and_run(
+            f"bcftools mpileup {cram} -f {self.ref} -T {regions} -a ad,format/dp --skip-indels -d 500 "
+            + f"| bcftools view -i 'AD[0:1] / format/DP >= {min_af}' -Oz -o {vcf}"
+        )
+        self.sp.print_and_run(f"bcftools index -t {vcf}")
+
+    def calc_hit_fraction(self, cram: str, ground_truth_bed: str) -> tuple[float, float, float]:
+        cram_base_name = os.path.basename(cram)
```
Comment (ilyasoifer) [inline ugvc/validation/variant_hit_fraction_caller.py]: I think it is much cleaner (and pure python) to read both (tiny) vcfs into dataframes (vcftools.get_vcf_df) and calculate intersection of indices

File: ugvc/pipelines/validation/quick_fingerprinting.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+import os
+
+from simppl.simple_pipeline import SimplePipeline
+
+from ugvc.utils.cloud_sync import cloud_sync
+from ugvc.validation.variant_hit_fraction_caller import VariantHitFractionCaller
+
+
+def __get_parser() -> argparse.ArgumentParser:
+    parser = argparse.ArgumentParser(prog="training_set_consistency_check", description=run.__doc__)
```
Comment (ilyasoifer) [inline ugvc/pipelines/validation/quick_fingerprinting.py:14]: Is this the right program name?

File: ugvc/pipelines/validation/quick_fingerprinting.py
Comment (ilyasoifer) [inline ugvc/pipelines/validation/quick_fingerprinting.py:1]: In terms of `pip install -e` inner works, would it be possible to split quick_fingerprinting.py into a really tiny script that only contains the "run" and a larger script `quick_fingerprinting_main.py` that contains all the logic (and is imported from `quick_fingerprinting.py`)? 
The reason is that `pip install -e `  copies the executable script and links to the imports, so changes in the imports will be reflected, but changes in the executable script will require new `pip install`

**Review by ilyasoifer — APPROVED** [review]

**Pull request merged**